### ISSUE TITLE
Binned x/y encoding ignores scale domain parameter

### ISSUE THREAD

**Issue by jheer** [vega/vega-lite#3575]
When using a binned spatial encoding, custom scale domain settings seem to be ignored. The generated Vega pulls the scale domain from the data, not the provided custom domain.

Here is an example spec:
```json
{
  "$schema": "https://vega.github.io/schema/vega-lite/v2.json",
  "data": {"values": [0.1, 0.6, 2.3]},
  "mark": "bar",
  "encoding": {
    "x": {
      "field": "data",
      "type": "quantitative",
      "bin": {"extent": [0, 10], "step": 0.5},
      "scale": {"domain": [0, 10]}
    },
    "y": {
      "aggregate": "count",
      "type": "quantitative",
      "scale": {"domain": [0, 2]}
    }
  }
}
```

I expect the x-axis domain to be `[0, 10]`, but instead it is `[0, 2.5]`.

**Comment by kanitw** [comment]
@yhoonkim  I remember you work on this. 

Can you remember why setting bin scale's domain is prevented?  

https://github.com/vega/vega-lite/blob/master/src/compile/scale/domain.ts#L177

(It should be relatively easy to fix, but wanna understand if fixing doesn't cause other bad effects / regression.)

**Comment by yhoonkim** [comment]
@kanitw 

I guess we prefer to control the scale by `bin.extent` instead of `scale.domain`. 

>But it's kinda weird that one of the bin parameter is outside of the bin object. (It's also more than just domain as it affects the bin transform that we perform before the scale.)
(https://github.com/vega/vega-lite/issues/2791#issuecomment-323247255)


> However,
it is still a problem when scale.domain and bin.extent are specified together.
>
>My suggestion is ignoring scale.domain when the field is binned and throwing a warning message.
(https://github.com/vega/vega-lite/pull/2429)

So we ignored `scale.domain`  and force to use `bin.extent` with a warning message. + We decided to fix it later...


--

The above Jeff's spec generates `x` scale in vega spec like the below: 
```
{
      "name": "x",
      "type": "linear",
      "domain": {
        "data": "data_0",
        "fields": [
          "bin_extent_0_10_step_0_5_data",
          "bin_extent_0_10_step_0_5_data_end"
        ]
      },
      "range": [
        0,
        {
          "signal": "width"
        }
      ],
      "zero": false
    }
```
However, `bin_extent_0_10_step_0_5_data_end` is 2.5 so it makes the x-domain  `[0,2.5]`. (It is also related to a standard aggregation behavior [that] (https://github.com/vega/vega/issues/983).) Thus, I think we have to keep using `bin.extent` as the domain but set vega's x-domain to be the user defined value (`[0, 10]`) instead of the value from the calculation (`[0, 2.5`).

**Comment by kanitw** [comment]
@yhoonkim can you help fix this then?

**Comment by yhoonkim** [comment]
@kanitw  I'll try! To solve this, I need to hear your thoughts about the below specs and their outcomes. Could you tell me if they look OK?

1) Spec
```json
{
    "$schema": "https://vega.github.io/schema/vega-lite/v2.json",
    "data": {"values": [0.1, 0.6, 2.3]},
    "mark": "bar",
    "encoding": {
      "x": {
        "field": "data",
        "type": "quantitative",
        "bin": {"extent": [0, 10], "step": 2},
        "scale": {"domain": [-1, 3]}
      },
      "y": {
        "aggregate": "count",
        "type": "quantitative",
        "scale": {"domain": [0, 2]}
      }
    }
  }
```

![image](https://user-images.githubusercontent.com/6269456/39091845-1a971c90-45b3-11e8-807c-dc31df8bf074.png)


2.
```json
{
    "$schema": "https://vega.github.io/schema/vega-lite/v2.json",
    "data": {"values": [0.1, 0.6, 2.3]},
    "mark": "bar",
    "encoding": {
      "x": {
        "field": "data",
        "type": "quantitative",
        "bin": {"extent": [0, 10], "step": 2},
        "scale": {"domain": [-10, 20]}
      },
      "y": {
        "aggregate": "count",
        "type": "quantitative",
        "scale": {"domain": [0, 2]}
      }
    }
  }
```
![image](https://user-images.githubusercontent.com/6269456/39091851-27fe2f36-45b3-11e8-8d11-fdb709831bd9.png)

**Comment by kanitw** [comment]
1) is ok

2) is a bit weird that the axis values is only drawn from the bin value.   In a way, if users customize `scale.domain`, then this is okay. 

The problem is that axis's `"values"` is set to 
```
"values": {
        "signal": "sequence(bin_extent_0_10_step_2_data_bins.start, bin_extent_0_10_step_2_data_bins.stop + bin_extent_0_10_step_2_data_bins.step, bin_extent_0_10_step_2_data_bins.step)"
      },
```

I think it's better if the start and stop are replaced with domain min and domain max. 

![image](https://user-images.githubusercontent.com/111269/39277990-9716474c-48a5-11e8-922a-595a4799f00e.png)

but I'm not sure how to do that in Vega expression and maybe it's okay to just drop `"values"` in such case.  

![image](https://user-images.githubusercontent.com/111269/39278013-af82044c-48a5-11e8-819d-8df9816634aa.png)


3) **More importantly**, this reminds me that users shouldn't have to set `domain` in addition to extent.

We should make scale domain depends on explicitly set `bin`'s `extent` too.   (This is even more important than correcting axis's "values".)

Basically,
```
{
    "$schema": "https://vega.github.io/schema/vega-lite/v2.json",
    "data": {"values": [0.1, 0.6, 2.3]},
    "mark": "bar",
    "encoding": {
      "x": {
        "field": "data",
        "type": "quantitative",
        "bin": {"extent": [-10, 10], "step": 2}
      },
      "y": {
        "aggregate": "count",
        "type": "quantitative",
        "scale": {"domain": [0, 2]}
      }
    }
  }
```

should have the domain = `[-10, 10]` nicely like this: 

![image](https://user-images.githubusercontent.com/111269/39277889-faf18c14-48a4-11e8-862b-6f4d9f05b9fc.png)

(I guess this isn't too magical and should be a reasonable default?)
